cpp/src/round_robin/round_robin.cu

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/device_vector.h>
+#include <thrust/scan.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <type_traits>
+#include <cmath> // for std::ceil()
+
+#include <cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/gather.cuh>
+
+namespace cudf {
+namespace experimental { 
+namespace detail {
+
+std::pair<std::unique_ptr<table>,
+          std::vector<cudf::size_type>>
+round_robin_partition(table_view const& input,
+                      cudf::size_type num_partitions,
+                      cudf::size_type start_partition = 0,
+                      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                      cudaStream_t stream = 0)
+{
+  auto nrows = input.num_rows();
+
+  CUDF_EXPECTS( num_partitions > 1 && num_partitions < nrows, "Incorrect number of partitions. Must be greater than 1 and less than number of rows." );
```

Comment:
It would be nice if the partitioner handled num_partitions >= nrows.  In the case where num_partitions == nrows, each partition has one row.  In the case where num_partitions > nrows, some partitions contain zero rows. I realize these are sort of degenerate cases, but let me explain the use-case where it can make sense from the caller's perspective.  Apologies for the wall of text.

Partitioning is always followed by a split call in preparation for transferring the partitions to different downstream tasks.  Round robin partitioning is just another partitioning algorithm like hash partitioning.  Arguably round-robin partitioning is just another form of hash partitioning where the hash function is the row number plus a specified offset.  Hash partitioning supports num_partitions >= nrows, which is why I'd like to see it supported here. That way the caller can always partition followed by a call to split and it works, regardless of the partitioning algorithm used.  Otherwise the caller needs to check for the round-robin num_partitions >= nrows case and handle splitting differently to avoid the code exploding.

Regarding how this would work in practice, here's an example for num_partitions == nrows:
With input table => col 1 {0, ..., 10}, num_partitions = 11, and start_partition = 2, the result should be:
table => col 1 {9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8}
offsets => {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
The table needs to be transformed when start_partition > 0, but the offsets are just a counting vector.

For the num_partitions > nrows case, take the same example with num_partitions=15.  We start handing out row 0 to partition 2, row 1 to partition 3, etc.  Eventually we run out of rows after partition 12, so partitions 13, 14, 0, and 1 would be empty.  This would be represented by the following output:
table => col 1 { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }
offsets => {0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 11 }
The output table is just a copy of the input, and the offsets vector is a slight twist on the counting vector, with the start offset and end offsets potentially duplicated some number of times.

These transforms could be handled by the caller, possibly more efficiently if the offsets don't need to be manifested and the table copied.  However it simplifies the caller logic if the data can _always_ be partitioned and split without needing to check the specific partition algorithm used and boundary cases of that algorithm.

I'm fine with this functionality being discussed/addressed in a followup issue.  If this is deferred, the limitation should be documented in the function declaration comments.

Reply:
1. If `num_partitions` can be larger than `nrows` then `start_partition` also can be larger than `num_rows`. I'm not sure how that makes sense.

2. Burn-down starts tomorrow. These are new specs requested 1 day prior to burn-down. They don't seem trivial so I doubt I can squeeze them in a day. So, the question is: do we push ahead with current implementation into `0.12` leaving additional requirements for the next release; or, do we insist on having the new specs before merging `round-robin`, which, most likely, would mean pushing this feature into a later release (moving back the PR from REVIEW to WIP, etc.)?

Again, specs should be discussed prior to REVIEW, in fact before the work even starts (WIP). So, if there are even more specs to this feature, let's expose everything _now_ so that we have a complete set of specs before implementation can move on. Please.

Reply:
> If num_partitions can be larger than nrows then start_partition also can be larger than num_rows. I'm not sure how that makes sense

It would work like this:

```
input: {7} // num_rows == 1
num_partitions: 3
start_partition: 2 // start_partition > num_rows

output: {7}
offsets: {0,0,0,1}
```

> Burn-down starts tomorrow.

Burn down is not the same as code freeze. Changes can continue to be made until code freeze. Don't let concerns about the release interfere with development.

> Again, specs should be discussed prior to REVIEW, in fact before the work even starts

In a perfect world, sure. But it's almost impossible to enumerate all requirements _a priori_. We like to get as many requirements as we can ahead of time, but ultimately design is an inherently iterative process.

Reply:
I apologize again. I missed the case where num_partitions > nrows in the examples.  I mistakenly assumed this would be handled similarly to the existing hash partition, and didn't occur to me to specify it in the example.  That is my fault.  I only noticed today because during review I saw the unit tests verifying the code explodes for this case.  I understand the frustration, and I truly am sorry for the misunderstanding and lack of full specification on my part.

> `start_partition` also can be larger than `num_rows`. I'm not sure how that makes sense.

`start_partition` specifies the starting partition, not row, of the round-robin algorithm.
For example, input table => col 1 {0, ..., 10}, num_partitions = 15, and start_partition = 14 would result in:
table => col 1 {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0}
offsets => {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 11, 11, 11}

This indicates partitions 10, 11, 12, and 13 are empty since start offset == end offset for those partitions. Whenever num_partitions >= nrows, the table output will be some rotation of the input table, and the offsets vector will be a counting vector with between 0 and 2 strings of duplicate entries to mark the ranges of empty partitions.

> do we push ahead with current implementation into 0.12 leaving additional requirements for the next release

I'll defer to the libcudf leads to what they prefer.  I don't see issues with integrating and releasing what is here now, since the future behavior change will only be in areas that currently throw. The Spark plugin will need to wait until this is addressed or find a way to work around the num_partitions >= nrows case to leverage this partitioning.  Not everything Spark needs is going to be in 0.12, so we're going to have to work with libcudf 0.13 snapshots even after 0.12 releases regardless.

> if there are even more specs to this feature, let's expose everything now so that we have a complete set of specs before implementation can move on.

I didn't explicitly call out null handling in the example, but the understanding is that the output validity vectors are arranged in a corresponding manner to the data vectors.
I am not aware of anything left unspecified at this point, but clearly I cannot be trusted in this regard. :sweat_smile:

Reply:
Null handling is tested in the tests. It has been from the very first release of the tests.

Reply:
@jlowe ,the cases `num_partitions == nrows`  and `num_partitions > nrows` seem to have very different logic than `num_partitions < nrows` and than each-other, in fact. 

I cannot reuse existent code (`num_partitions < nrows`) for either of the 2 special cases above. Because `num_partitions >= nrows`

   implies:
   `max_partition_size = ceil( nrows / num_partitions ) == 1;`
   
   which implies:
   `index_within_partition = (...) % (max_partition_size-1)`: division by 0;
   `partition_index        = (...) / (max_partition_size-1)`: division by 0;

That means the code must branch-out into 2 additional separate cases. I think I have a quick solution for `num_partitions == nrows`. However, the case `num_partitions > nrows` seems more involved (frankly, I don't understand it, yet).

Reply:
@jrhemstad, @jlowe are you against merging this PR and then opening a new one (enhancement) to continue work with the new specs?

Reply:
num_partitions > nrows is handled by generating empty partitions.

Think of the algorithm as dealing out cards.
- the deck of cards is represented as the rows in the table.
- the number of partitions is the number of players being dealt cards.
- the start_partition indicates which player starts getting cards first.

The algorithm result is two things:
- Another deck of cards formed by stacking each player's cards back into a deck again, preserving the order of cards dealt to each player, starting with player 0.
- A vector into the output deck indicating where a player's cards start.

A player's deck (partition) is the range of cards starting at the corresponding offset and ending at the next player's starting offset or the last card in the deck if it's the last player.

When num_partitions > nrows, we have more players than decks.  We start dealing to the first indicated player and continuing around the players until we run out of cards before we run out of players.  Players that did not get any cards are represented by
`offset[i] == offset[i+1]` or
`offset[i] == table.num_rows()` if `i == num_partitions-1`
meaning there are no cards (rows) in their deck (partition).

Hopefully with that analogy the examples I gave earlier for the case of num_partitions > nrows will make sense.

Reply:
I'm cool with it going in and addressing this as a followup.  I assume @jrhemstad is as well given he approved the PR after commenting on this issue thread.